improver/calibration/rainforest_calibration.py

```diff
@@ -461,124 +465,123 @@ def _calculate_error_probabilities(
+            probability_CDF:
+                Cube containing the CDF of probabilities for each ensemble member at model
                 thresholds.
             forecast:
                 Cube containing NWP ensemble forecast.
             output_thresholds:
-                Ordered list of thresholds at which to calculate the output probabilities.
+                Sorted array of thresholds at which to calculate the output probabilities.
 
         Returns:
             Cube containing probabilities at output thresholds for all realizations. Dimensions
             are same as forecast cube with additional threshold dimension first.
         """
 
-        # Add error to forecast. The thresholds for each grid point are obtained
-        # by adding the forecast at that point to error_values; thus
-        # we get a different set of thresholds for each grid point.
-        error_values = error_CDF.coord(var_name="threshold").points
-        error_values = np.expand_dims(
-            error_values, axis=tuple(range(1, forecast.ndim + 1))
-        )
-        thresholds = error_values + forecast.data[np.newaxis, :]
-        probabilities = error_CDF.data
-
-        # Add an extra threshold below/above with one/zero probability to
-        # ensure that the PDF covers the full probability range.
-        # Thresholds are usually float32 with epsilon of ~= 1.1e-7
-        eps = np.finfo(thresholds.dtype).eps
-        # For small values (especially exactly zero), at least epsilon;
-        # for larger values, the next representable float number.
-        thresholds = np.concatenate(
-            [
-                thresholds[[0]] - np.maximum(eps, np.abs(thresholds[[0]] * eps)),
-                thresholds,
-                thresholds[[-1]] + np.maximum(eps, np.abs(thresholds[[-1]] * eps)),
-            ],
-            axis=0,
-        )
-        probabilities = np.concatenate(
-            [
-                np.ones((1,) + probabilities.shape[1:], np.float32),
-                probabilities,
-                np.zeros((1,) + probabilities.shape[1:], np.float32),
-            ],
-            axis=0,
-            dtype=np.float32,
-        )
-
-        # get bounds
+        input_probabilties = probability_CDF.data
+        output_thresholds = np.array(output_thresholds, dtype=np.float32)
         bounds_data = get_bounds_of_distribution(forecast.name(), forecast.units)
-        lower_bound_in_fcst_units = bounds_data[0]
-
-        # Set probability to 1 for thresholds <= lower bound
-        threshold_below_bound = thresholds <= lower_bound_in_fcst_units
-        thresholds = np.where(
-            threshold_below_bound, lower_bound_in_fcst_units, thresholds
-        )
-        probabilities = np.where(threshold_below_bound, 1, probabilities)
+        lower_bound = bounds_data[0].astype(np.float32)
+        if (len(self.model_thresholds) == len(output_thresholds)) and np.allclose(
+            self.model_thresholds, output_thresholds
+        ):
```

Comment:
I prefer the current approach because it  allows the case where the model thresholds are the same as the output thresholds but in a different unit. In this case, `process` will convert the output thresholds to the units of the forecast, then pass them into `self._get_ensemble_distributions`, which will recognise that they are now the same as `self.model_thresholds`, and therefore the interpolation step can be skipped.

Reply:
True, but this alternative avoids having to specify the thresholds at all (and so avoids the question of units). I guess the advantage of specifying the output thresholds is that you know exactly what you are going to get.

Reply:
Yeah I think forcing the user to be specific in this way doesn't hurt, and it means that the user doesn't need to check what the model thresholds are, they can just pass in their desired output thresholds and the code will automatically skip the interpolation if possible.